### ISSUE TITLE
[QA] Verify FlashAttention + Liger Kernel + Unsloth speedups

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#78]
## What needs verifying

Three orthogonal speedup paths, all gated behind config flags:

1. `use_flash_attn: true` + `flash_attn_version: 2` (or `3` on Hopper+)
2. `use_liger: true` (Liger Kernel fused ops: RMSNorm / SwiGLU / RoPE / cross-entropy)
3. `backend: unsloth` (FastLanguageModel — claims 2-5× speedup on a single GPU)

## Hardware / software prerequisites

- NVIDIA GPU with ≥ 12 GB VRAM (Ampere+ for FlashAttention 2; Hopper / Blackwell for FA3)
- `pip install flash-attn --no-build-isolation` (FA2)
- `pip install 'soup-cli[liger]'` (Liger Kernel)
- `pip install 'soup-cli[fast]'` (Unsloth)
- A baseline single-GPU `soup train` run to compare against

## Steps

```bash
# Baseline
soup train --config soup.yaml --output ./baseline

# +FA
edit soup.yaml: training.use_flash_attn: true
soup train --config soup.yaml --output ./fa

# +Liger
edit soup.yaml: training.use_liger: true
soup train --config soup.yaml --output ./liger

# Unsloth (different backend, fresh config)
soup recipes use llama-3.2-1b-sft-unsloth
soup train --config soup.yaml --output ./unsloth
```

Compare loss curves + steps/sec between the four runs.

## Success

- All four runs converge to similar loss
- FlashAttention: ≥ 1.3× faster than baseline on the same GPU
- Liger: ≥ 1.1× faster, with lower peak VRAM
- Unsloth: ≥ 1.5× faster than transformers backend (often 2-3×)
- No silent degradation: `soup chat --model ./fa` produces coherent output

## Report back

- `soup version --full`, `pip show flash-attn liger-kernel unsloth`
- GPU model + driver + CUDA
- Per run: steps/sec, peak VRAM, final loss
- Any "falling back to slow path" warnings

## If it fails

Open separate bug per failing component. Note: FlashAttention 3 is Hopper+ only — don't file a bug if FA3 falls back to FA2 on Ampere.

**Comment by MakazhanAlpamys** [comment]
## Measured for the first time — on an H100, since neither kernel installs on Windows

Real `soup train`, Llama-3.1-8B + LoRA r=16, bf16, batch 4, seq 1024, 80 steps, 5 repeats **interleaved** A/B/C/D so clock drift cannot favour whichever ran first.

| arm | median tok/s | spread | peak VRAM | SM clock (busy) | ratio |
|---|---|---|---|---|---|
| baseline (SDPA) | 9621.1 | 0.83% | 70.99 GiB | 1965 MHz | 1.000x |
| FlashAttention 2 | 9764.8 | 0.92% | 70.99 GiB | 1965 MHz | **1.015x** |
| Liger | 10115.3 | 0.22% | 61.81 GiB | 1905 MHz | **1.051x**, VRAM −12.9% |
| FA2 + Liger | 10256.0 | 0.53% | 61.81 GiB | 1920 MHz | **1.066x**, VRAM −12.9% |

Activation was **verified, not assumed** — a call counter in dedicated short runs: baseline 256 SDPA / 0 FA2; FA2 256 FA2 / **0 SDPA**; Liger with module classes actually swapped to `LigerRMSNorm` / `LigerSwiGLUMLP`. Baseline's real kernel from `torch.profiler` is `cudnn_generated_fort_native_sdpa_sm90_flash_fprop_wgmma_f16`.

### The documented claims were wrong, and are now corrected

- FlashAttention: **"2-4x speedup and significant memory savings"** → measured **1.015x, zero memory saving**
- Liger: **"20-60% memory and 20-40% throughput"** → measured **12.9% and 5.1%**

Both corrected in the source docstrings and `docs/`, with the measurement conditions attached. FA2 has little to beat here because torch SDPA already dispatches a Hopper flash kernel — that is a property of the comparison, not a criticism of FA2, and the gap would widen with sequence length.

### The benchmark found bugs before it found numbers

1. **`training.use_liger: true` crashed at step 0** — Soup patched the model but never set `TrainingArguments.use_liger_kernel`, so TRL's entropy guard could not fire and `entropy_from_logits(None)` raised. Reproduced on trl 0.26.2 **and** 0.19.1 — the feature had never worked anywhere in the supported pin. **Fixed** (`38c3753`).
2. **`data.max_length` above 1024 was silently ignored on every SFT run** (`4096` → 1024 tokens/sample). **Fixed** (`38c3753`). This also bounds the FA result: 1024 was the longest sequence Soup could produce, so this *is* the operating point users got.
3. **FA3 selection is unreachable dead code** — filed as #334.

### Not measurable, and why
**FlashAttention 3**: no `nvcc` on the box to build `hopper/`, PyPI `flash-attn-3` is a 0.0.0 placeholder, and the `kernels` route needs `huggingface_hub>=1.10` against transformers 4.57.6's `<1.0` pin (tried, broke transformers, reverted). And even installed, Soup could not select it (#334). So FA3 stays **unmeasured**, and "Hopper now or never" did not rescue it.

### Caveat
One point in config space: 8B + LoRA + seq 1024 + batch 4, no gradient checkpointing. Liger's memory saving would shrink with gradient checkpointing on. Nothing was tuned to improve any arm.

Record: `benchmarks/gate-h100-validation.md` STEP 17.

**Comment by MakazhanAlpamys** [comment]
Measured, and the claims corrected. CI green (run 31189778999 carries the fixes; the fix commit's own run was red on an unrelated CPU tolerance, since corrected).

Llama-3.1-8B + LoRA, bf16, batch 4, seq 1024, 5 repeats **interleaved** A/B/C/D so clock drift cannot favour whichever ran first. SM clock 1965 MHz busy.

| arm | median tok/s | spread | peak VRAM | ratio |
|---|---|---|---|---|
| baseline (SDPA) | 9621.1 | 0.83% | 70.99 GiB | 1.000x |
| FlashAttention 2 | 9764.8 | 0.92% | 70.99 GiB | **1.015x** |
| Liger | 10115.3 | 0.22% | 61.81 GiB | **1.051x**, −12.9% |
| FA2 + Liger | 10256.0 | 0.53% | 61.81 GiB | 1.066x |

Activation **verified**, not assumed: baseline 256 SDPA / 0 FA2; FA2 256 FA2 / **0 SDPA**; Liger with module classes actually swapped. Baseline's real kernel is `cudnn_generated_fort_native_sdpa_sm90_flash_fprop_wgmma_f16` — already a Hopper flash kernel, which is why FA2 has little to beat and why the VRAM columns match to four figures.

**The documented claims were wrong and are corrected in the source and docs:** FA "2-4x + significant memory savings" → 1.015x, zero memory saving. Liger "20-60% memory / 20-40% throughput" → 12.9% / 5.1%.

Two bugs fixed on the way (`38c3753`):
- `training.use_liger: true` **crashed at step 0** across the whole supported trl pin — the feature had never run
- `data.max_length` above 1024 was **silently ignored on every SFT run** (4096 → 1024 tokens/sample)

Try:
```bash
soup train --config soup.yaml   # training.use_liger: true, training.use_flash_attn: true
```

**FlashAttention 3 remains unmeasured** and is tracked in #334 — Soup's detection cannot fire at all (wrong package name), and independently there was no way to build it on this box. "Hopper now or never" did not rescue it, and saying so is more useful than a blank.

Caveat: one operating point (8B + LoRA + seq 1024 + batch 4, no gradient checkpointing). Liger's saving shrinks with checkpointing on; FA's gain grows with sequence length — and 1024 was the longest sequence Soup could produce, which was itself the bug fixed here. Nothing was tuned to improve any arm.

Record: `benchmarks/gate-h100-validation.md` STEP 17.
